### ISSUE TITLE
Monitor consensus health of the devnet

### ISSUE THREAD

**Issue by geigerzaehler** [oscoin/oscoin#604]
To verify that consensus in the devnet works we want to check that

* All nodes agree on a blockchain
* Average block production rate is sensible.

We need to come up with tools that allow us to verify this. Ideally this information would be available on Grafana but a CLI tool would do, too.

**Comment by adinapoli-mndc** [comment]
I have had a brief discussion with @kim and after evaluating a bunch of options (or lack thereof) let me try to start with a straw-man proposal. Unfortunately, our original plan of using prometheus & EKG to simply record the tip hash as a label got quickly destroyed when we realised that [prometheus doesn't support that](https://github.com/prometheus/prometheus/issues/2227). We cannot fallback on a schema where we simply log the height of a node as a `gauge` and we attach a label to it (i.e, the tip hash) because Prometheus' data storage is such that it will treat this new entry `(height,hash)` as a new timeserie, generating a new key in the K/V store and eventually saturating the memory, crashing the process.

Kim's idea was to use one of the many Grafana's plugins to get the job done, and specifically piggyback of the [stackdriver](https://grafana.com/docs/features/datasources/stackdriver) one. Unfortunately, this plugin doesn't support the `logging` service of stackdriver, which means we cannot curate the logs and transform them into metrics. What this plugin support though is the [monitoring](https://cloud.google.com/monitoring) service.  

Now, `stackdriver monitoring` doesn't support a textual label as a first-class metric type, but it does support gauges. So, here is my straw-man proposal:

* We write a [custom metric type](https://cloud.google.com/monitoring/agent/custom-metrics-agent) for the `stackdriver monitor` agent by extending our devops scripts to instrument [collectd](https://collectd.org/)  correctly;

* We treat this new metric type as a gauge with a label attached; essentially the same we would have done for Prometheus with the hope that Google's engineer got the data storage right;

* We would use the Grafana Stackdriver plugin to aggregate the gauge by using an [alignment](https://cloud.google.com/monitoring/charts/metrics-selector#alignment) of 1minute, so we should hopefully see every minute a bunch of dots plotting the height of each node, and the tipHash associated with it;

* I don't know `collectd` at all and I have just skimmed through it, but it seems like it can be instrumented to simply poll an HTTP endpoint, so we could expose a `Middleware` (in the `wai` sense) to make this information available.

## Advantages

* Application agnostic;
* Easy to implement application side (I think we can bypass the metrics system completely and simply have a `Middleware` call `getTip` and return the result);

## Disadvantages

* Devops pain.

**Comment by kim** [comment]
@adinapoli-mndc Yeah this _might_ work. It's not clear though if COS supports the "normal" stackdriver monitoring agent (aka collectd), although there exists a configuration at `/etc/stackdriver/monitoring.config.d/collectd.conf`.

**Comment by adinapoli-mndc** [comment]
> It's not clear though if COS supports the "normal" stackdriver monitoring agent (aka collectd), although there exists a configuration at /etc/stackdriver/monitoring.config.d/collectd.conf.

Right, that's worth investigating. Alternatively, the stackdriver monitor configuration says you can also push via `statsd` (and there are Haskell libraries) but I have no clue how much work that is to instrument the `monitoring` agent to read via `statsd` vs `collectd`.
